File: src/ipa/raspberrypi/controller/rpi/black_level.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/*
 * Copyright (C) 2019, Raspberry Pi Ltd
 *
 * black_level.cpp - black level control algorithm
 */

#include <math.h>
#include <stdint.h>

#include <libcamera/base/log.h>

#include "../black_level_status.h"

#include "black_level.h"

using namespace RPiController;
using namespace libcamera;

LOG_DEFINE_CATEGORY(RPiBlackLevel)

#define NAME "rpi.black_level"

BlackLevel::BlackLevel(Controller *controller)
	: Algorithm(controller)
{
}

char const *BlackLevel::name() const
{
	return NAME;
}

int BlackLevel::read(boost::property_tree::ptree const &params)
{
	uint16_t blackLevel = params.get<uint16_t>(
		"black_level", 4096); /* 64 in 10 bits scaled to 16 bits */
	blackLevelR_ = params.get<uint16_t>("black_level_r", blackLevel);
	blackLevelG_ = params.get<uint16_t>("black_level_g", blackLevel);
	blackLevelB_ = params.get<uint16_t>("black_level_b", blackLevel);
	LOG(RPiBlackLevel, Debug)
		<< " Read black levels red " << blackLevelR_
		<< " green " << blackLevelG_
		<< " blue " << blackLevelB_;
	return 0;
}

void BlackLevel::prepare(Metadata *imageMetadata)
{
	/*
	 * Possibly we should think about doing this in a switchMode or
	 * something?
	 */
	struct BlackLevelStatus status;
	status.blackLevelR = blackLevelR_;
	status.blackLevelG = blackLevelG_;
	status.blackLevelB = blackLevelB_;
	imageMetadata->set("black_level.status", status);
}

/* Register algorithm with the system. */
static Algorithm *create(Controller *controller)
{
	return new BlackLevel(controller);
}
static RegisterAlgorithm reg(NAME, &create);

File: src/ipa/rpi/controller/rpi/black_level.zig
const libcamera = @import("libcamera");
const Metadata = @import("../metadata.zig").Metadata;
const BlackLevelStatus = @import("../black_level_status.zig").BlackLevelStatus;
const Algorithm = @import("../algorithm.zig").Algorithm;

const NAME = "rpi.black_level";

pub const BlackLevel = struct {
    algorithm: Algorithm,
    blackLevelR: f64,
    blackLevelG: f64,
    blackLevelB: f64,

    pub fn init(controller: *Controller) BlackLevel {
        return BlackLevel{
            .algorithm = Algorithm.init(controller),
            .blackLevelR = 0.0,
            .blackLevelG = 0.0,
            .blackLevelB = 0.0,
        };
    }

    pub fn name(self: *BlackLevel) []const u8 {
        return NAME;
    }

    pub fn read(self: *BlackLevel, params: YamlObject) i32 {
        const blackLevel = params.get("black_level").getOptional(u16, 4096);
        self.blackLevelR = params.get("black_level_r").getOptional(u16, blackLevel);
        self.blackLevelG = params.get("black_level_g").getOptional(u16, blackLevel);
        self.blackLevelB = params.get("black_level_b").getOptional(u16, blackLevel);
        log.debug("Read black levels red {f} green {f} blue {f}", .{ self.blackLevelR, self.blackLevelG, self.blackLevelB });
        return 0;
    }

    pub fn initialValues(self: *BlackLevel, blackLevelR: *u16, blackLevelG: *u16, blackLevelB: *u16) void {
        blackLevelR.* = @intCast(u16, self.blackLevelR);
        blackLevelG.* = @intCast(u16, self.blackLevelG);
        blackLevelB.* = @intCast(u16, self.blackLevelB);
    }

    pub fn prepare(self: *BlackLevel, imageMetadata: *Metadata) void {
        var status = BlackLevelStatus{
            .blackLevelR = @intCast(u16, self.blackLevelR),
            .blackLevelG = @intCast(u16, self.blackLevelG),
            .blackLevelB = @intCast(u16, self.blackLevelB),
        };
        imageMetadata.set("black_level.status", status);
    }
};

pub fn create(controller: *Controller) *BlackLevel {
    return BlackLevel.init(controller);
}

pub const RegisterAlgorithm = struct {
    name: []const u8,
    createFunc: fn(*Controller) *BlackLevel,

    pub fn init(name: []const u8, createFunc: fn(*Controller) *BlackLevel) RegisterAlgorithm {
        return RegisterAlgorithm{ .name = name, .createFunc = createFunc };
    }
};

pub const reg = RegisterAlgorithm.init(NAME, create);
